[클린코드 3기 이현정] 자동차 경주 미션 STEP 1

jlee0505 opened the pull request:
# 🎯 1단계 요구사항
✅ 자동차에 이름을 부여할 수 있다. 전진하는 자동차를 출력할 때 자동차 이름을 같이 출력한다.
✅ 자동차 이름은 쉼표(,)를 기준으로 구분하며 이름은 5자 이하만 가능하다.
✅ 사용자는 몇 번의 이동을 할 것인지를 입력할 수 있어야 한다.
✅ 주어진 횟수 동안 n대의 자동차는 전진 또는 멈출 수 있다.
✅ 전진하는 조건은 0에서 9 사이에서 random 값을 구한 후 random 값이 4 이상일 경우 전진하고, 3 이하의 값이면 멈춘다.
<hr/>
# 1단계 1차 pr (7/25)
## ✋ 질문
: 유조님과 페어프로그래밍을 진행하면서 자동차 이름과 시도횟수 값을 받아올 때 상태관리를 html 이 아닌 앱안에서 할 수 있도록 모델을 만들어보라고 리뷰를 받았었는데 잘 이해가 가지 않았습니다. 혹시 추가 설명이나 관련 키워드 알 수 있을까요?

## 🔖 추신
마지막 테스트 케이스를 완성하지 못했습니다 ㅜㅜ 마저 해서 피드백 반영 때 함께 제출하도록 하겠습니다!
<hr/>
# 1단계 2차 pr (7/27)

리뷰어님이 유조님이신줄 몰랐어요.. 유조님의 코드를 유조님께 리뷰해달라 요청한 것 같아 괜히 민망했던...ㅎㅎ

## ✋ 질문

'constant라는 이름을 가진 모듈에는 상수들이 존재할거라고 유추되고, utils라는 이름을 가진 모듈에는 범용적으로 쓰일 수 있는 util 함수들이 존재할거라고 유추 되는 것처럼 현재 나눠놓으신 모듈들에 알맞게 함수나 변수가 들어있는지 한번 점검해보시면 좋을거 같다는 생각을 했습니다!' 

=> 1. 이 부분에 대해서 기존에 index.js 에 있던 handle~ 함수들을 handlers.js 라는 모듈로 따로 빼서 정리하고자 했는데 어떻게 할 수 있을지 궁리해보았으나, 그럼 index.js 로 앱이 시작할 때 handler 들을 실행시킬 수 있도록 어떻게 연결하는지 찾지 못했습니다. 

=> 2. 그외에 다른 부분을 의미하신 건지, 있다면 알려주시면 감사하겠습니다!

## 🔖 추신
상태 관리에 대한 부분은 공부중인데 아직 반영하진 못했습니다. 일단 2단계 요구사항 진행하며 같이 고민해보겠습니다!

yujo11 commented on src/template.js:
```diff
@@ -0,0 +1,25 @@
+import { $ } from "./dom.js";
+import { SELECTORS } from "./constants.js";
+
+export const templateSpinner = `<div class="spinners d-flex justify-center mt-3">
+                                  <div class="relative spinner-container">
+                                    <span class="material spinner"></span>
+                                  </div>
+                                </div>`;
+
+export const templateForward = `<div class="forward-icon mt-2">⬇️️</div>`;
+
+export const createTemplateCarPlayer = ($input) => {
+  const template = $input.value.split(",").map(
+    (name) => `<div class="car">
+                  <div class="car-player">${name}</div>
+                  <div class="spinners d-flex justify-center mt-3">
+                    <div class="relative spinner-container">
+                      <span class="material spinner"></span>
+                    </div>
+                  </div>
+                </div>`
+  );
+
+  $(SELECTORS.CAR_PLAYER_WRAPPER_DIV).innerHTML = template;
+};
```
`templateSpinner`와 `templateForward`는 단순히 템플릿을 저장하는 변수의 역할만 하고 있는데 `createTemplateCarPlayer`는 템플릿을 생성해서 DOM에 추가해주는 역할까지 하고 있네요.

`createTemplateCarPlayer`가 하는 두 가지 일을 다른 함수로 분리해보시면 어떨까요?

yujo11 commented on src/utils.js:
```diff
@@ -0,0 +1,3 @@
+export const displaySelector = ($selector) => {
+  $selector.classList.remove("hidden");
+};
```
`displaySelector`라는 이름은 `display`를 `select`하는 함수라고 생각이 될 수도 있을거 같아요

다른 이름으로 바꿔보는건 어떨까요?

yujo11 reviewed (changes requested):
안녕하세요 현정님! LMS에서 리뷰이가 안 보여서 현정님 리뷰어가 저인줄 몰랐네요ㅎㅎ...

요구사항을 구현하기 위해 필요한 코드들만 간결하게 있어서 전체적인 코드를 읽기가 좋았습니다!

다만 현재 다양한 모듈(파일)들이 존재하는데요. 각각의 모듈 안에 존재하는 함수나 변수들이 모듈이름과 걸맞게 존재하고 있는지는 조금 의문이 드는 부분이 있었습니다.

`constant`라는 이름을 가진 모듈에는 상수들이 존재할거라고 유추되고, `utils`라는 이름을 가진 모듈에는 범용적으로 쓰일 수 있는 util 함수들이 존재할거라고 유추 되는 것처럼 현재 나눠놓으신 모듈들에 알맞게 함수나 변수가 들어있는지 한번 점검해보시면 좋을거 같다는 생각을 했습니다!

>유조님과 페어프로그래밍을 진행하면서 자동차 이름과 시도횟수 값을 받아올 때 상태관리를 html 이 아닌 앱안에서 할 수 있도록 모델을 만들어보라고 리뷰를 받았었는데 잘 이해가 가지 않았습니다. 혹시 추가 설명이나 관련 키워드 알 수 있을까요?

이 부분은 반드시 모델을 사용해야 한다고 말씀드렸던 부분은 아니고, 모델을 만들어서 사용하는 것도 가능할거 같다는 취지였습니다.

현재 `startRaicingGames`라는 함수를 보면 아래와 같은데요.

```
export function startRacingGame(times) {
  let count = 1;
  const racingCars = document.querySelectorAll(SELECTORS.CAR_PLAYER_DIV);

  const timeoutId = setInterval(() => {
    // isFoward여부에 따라 레이싱 카들에 전진 템플릿을 삽입한다.

    racingCars.forEach((racingcar) => {
      if (isFoward()) {
        racingcar.insertAdjacentHTML("afterend", templateForward);
      }
    });

    if (count++ === times) {
      clearInterval(timeoutId);
      // 스피너를 제거하는 로직
    }
  }, 1000);
}
```

게임을 진행하기 위해서 DOM에 추가해둔 `CAR_PLAYER`들을 가져오고 있습니다.

이렇게 함수가 작성된 이유는 이전에 사용자에게 입력받은 자동차 이름들을 모두 DOM에 넣고 별도의 상태로 가지고 있지 않고,  자동차 경주를 진행하기 위해서 현정님아 작성하신 어플리케이션이 자동차경주를 진행하기 위해서는 다시 자동차들을 알아야하기 때문인데요.

이렇게 어플리케이션의 생명주기에서 지속적으로 필요한 "자동차"를 사용자로부터 입력받았을 때 DOM에 넣어버리는게 아니라 어플리케이션에서 "자동차"를 "상태"로 관리해보면 좋을거 같아서 말씀을 드렸던 부분입니다.

어플리케이션에서 상태를 관리하는 부분이 어렵게 느껴지실 수도 있겠으나 프론트엔드 개발에 있어 가장 중요한 개념 중 하나라고 생각되는만큼 이번 기회에 "상태(state)"가 무엇인지 "상태(state)"를 어떻게 관리해야하는지 고민해보고 적용해보셨으면 좋겠네요!

step1 진행하느라 고생 많으셨습니다! 👍 👍 👍

jlee0505 commented on src/index.js:
```diff
@@ -0,0 +1,44 @@
+  e.preventDefault();
+
+  // 각 자동차 이름이 5글자가 넘으면 경고 메세지를 표시한다.
+  if (e.target[1].value) {
```
음... 영 감이 안잡히는데 혹시 힌트라도 얻을 수 있을까요??

yujo11 commented on cypress/e2e/racing.cy.js:
```diff
@@ -1,5 +1,13 @@
 import { SELECTORS, ERROR_MESSAGES } from "/src/constants.js";
-import { shouldShowAlert } from "./utils.js";
+
+//Custom Commands
+Cypress.Commands.add("shouldShowAlert", (typeVal, selector, err) => {
+  cy.on("window:alert", (alertMessage) => {
+    expect(alertMessage).to.eq(err);
+  });
+  if (typeVal) cy.get(selector).type(typeVal);
+  cy.get(SELECTORS.CAR_NAME_FORM).submit();
+});
```
라이브러리를 사용할 때는 해당 라이브러리에서 제시하는 컨벤션이나 구조를 따라주면 좋은데요.

`cypress/support/command.js`로 옮기는건 어떨까요?

yujo11 reviewed (approved):
안녕하세요 현정님!

리뷰 꼼꼼하게 읽어보고 반영해주셔서 감사합니다!

>=> 1. 이 부분에 대해서 기존에 index.js 에 있던 handle~ 함수들을 handlers.js 라는 모듈로 따로 빼서 정?>리하고자 했는데 어떻게 할 수 있을지 궁리해보았으나, 그럼 index.js 로 앱이 시작할 때 handler 들을 실행시>킬 수 있도록 어떻게 연결하는지 찾지 못했습니다.
>=> 2. 그외에 다른 부분을 의미하신 건지, 있다면 알려주시면 감사하겠습니다!

자동차 경주 미션을 진행하시면서 여러 함수들과 파일들을 만드셨는데 각각의 함수들이 있을법한 위치에 있었으면 좋겠다는 생각으로 드린 리뷰였습니다!

예를 들어 `racing.js`라는 파일에 있는 `getRandomInt`라는 함수는 범용적으로 사용될 수 있으니 uitl에 있어도 되겠다는 식으로요!

각각의 파일마다 네이밍을 하는 이유는 해당 파일에서 어떤 역할을 하는 함수들을 가지고 있는지 명시적으로 표시하기 위함인데요. 함수들마다 적절한 파일(모듈)로 묶어서 작성한다면 자연스럽게 코드를 읽을 때 가독성도 좋아지는데요. 앞으로 코드를 작성하실 때도 이 점도 함께 고려해보시면 좋을거 같습니다!

>상태 관리에 대한 부분은 공부중인데 아직 반영하진 못했습니다. 일단 2단계 요구사항 진행하며 같이 고민해보겠습니다!

당장 이번 미션에 적용하지 못하더라도 고민해보고 찾아보시는 것만으로도 정말 대단하다고 생각합니다!


이번 step은 이만 머지하도록 하겠습니다! 다음 step에서 다시 뵙겠습니다! 미션 진행하느라 고생 많으셨습니다 현정님! 👍 👍 👍

jlee0505 commented:
블로그 글 정리한게 비공개로 되어있었네요 ㅠㅠ 여기 다시 열어두었습니다! 보시고 제가 이해한 게 맞는지 확인 한 번 해주시면
감사하겠습니다 🙇🏻‍♀️
***@***.***/%EC%83%81%ED%83%9C-%EA%B4%80%EB%A6%AC

감사합니다,

정

2022년 7월 25일 (월) 오후 11:18, JO YUN HO ***@***.***>님이 작성:

> ***@***.**** requested changes on this pull request.
>
> 안녕하세요 현정님! LMS에서 리뷰이가 안 보여서 현정님 리뷰어가 저인줄 몰랐네요ㅎㅎ...
>
> 요구사항을 구현하기 위해 필요한 코드들만 간결하게 있어서 전체적인 코드를 읽기가 좋았습니다!
>
> 다만 현재 다양한 모듈(파일)들이 존재하는데요. 각각의 모듈 안에 존재하는 함수나 변수들이 모듈이름과 걸맞게 존재하고 있는지는 조금
> 의문이 드는 부분이 있었습니다.
>
> constant라는 이름을 가진 모듈에는 상수들이 존재할거라고 유추되고, utils라는 이름을 가진 모듈에는 범용적으로 쓰일 수
> 있는 util 함수들이 존재할거라고 유추 되는 것처럼 현재 나눠놓으신 모듈들에 알맞게 함수나 변수가 들어있는지 한번 점검해보시면 좋을거
> 같다는 생각을 했습니다!
>
> 유조님과 페어프로그래밍을 진행하면서 자동차 이름과 시도횟수 값을 받아올 때 상태관리를 html 이 아닌 앱안에서 할 수 있도록 모델을
> 만들어보라고 리뷰를 받았었는데 잘 이해가 가지 않았습니다. 혹시 추가 설명이나 관련 키워드 알 수 있을까요?
>
> 이 부분은 반드시 모델을 사용해야 한다고 말씀드렸던 부분은 아니고, 모델을 만들어서 사용하는 것도 가능할거 같다는 취지였습니다.
>
> 현재 startRaicingGames라는 함수를 보면 아래와 같은데요.
>
> export function startRacingGame(times) {
>   let count = 1;
>   const racingCars = document.querySelectorAll(SELECTORS.CAR_PLAYER_DIV);
>
>   const timeoutId = setInterval(() => {
>     // isFoward여부에 따라 레이싱 카들에 전진 템플릿을 삽입한다.
>
>     racingCars.forEach((racingcar) => {
>       if (isFoward()) {
>         racingcar.insertAdjacentHTML("afterend", templateForward);
>       }
>     });
>
>     if (count++ === times) {
>       clearInterval(timeoutId);
>       // 스피너를 제거하는 로직
>     }
>   }, 1000);
> }
>
> 게임을 진행하기 위해서 DOM에 추가해둔 CAR_PLAYER들을 가져오고 있습니다.
>
> 이렇게 함수가 작성된 이유는 이전에 사용자에게 입력받은 자동차 이름들을 모두 DOM에 넣고 별도의 상태로 가지고 있지 않고, 자동차
> 경주를 진행하기 위해서 현정님아 작성하신 어플리케이션이 자동차경주를 진행하기 위해서는 다시 자동차들을 알아야하기 때문인데요.
>
> 이렇게 어플리케이션의 생명주기에서 지속적으로 필요한 "자동차"를 사용자로부터 입력받았을 때 DOM에 넣어버리는게 아니라
> 어플리케이션에서 "자동차"를 "상태"로 관리해보면 좋을거 같아서 말씀을 드렸던 부분입니다.
>
> 어플리케이션에서 상태를 관리하는 부분이 어렵게 느껴지실 수도 있겠으나 프론트엔드 개발에 있어 가장 중요한 개념 중 하나라고
> 생각되는만큼 이번 기회에 "상태(state)"가 무엇인지 "상태(state)"를 어떻게 관리해야하는지 고민해보고 적용해보셨으면 좋겠네요!
>
> step1 진행하느라 고생 많으셨습니다! 👍 👍 👍
> ------------------------------
>
> In cypress/e2e/racing.cy.js
> <https://github.com/next-step/js-racingcar/pull/101#discussion_r928828908>
> :
>
> > +describe("intro: 유저가 첫 화면을 정상적으로 볼 수 있는지 테스트한다.", () => {
> +  beforeEach(() => {
> +    cy.visit("http://127.0.0.1:5501/index.html");
> +  });
> +
> +  context("처음 시작했을 때", () => {
> +    it("자동차 이름 입력 외에 다른 창은 뜨지 않는다", () => {
> +      cy.get(SELECTORS.CAR_NAME_FORM).should("not.have.class", "hidden");
> +    });
> +  });
> +});
>
> 👍👍👍
> ------------------------------
>
> In cypress/e2e/utils.js
> <https://github.com/next-step/js-racingcar/pull/101#discussion_r928910882>
> :
>
> > +import { SELECTORS } from "/src/constants.js";
> +
> +export const shouldShowAlert = (typeVal, selector, err) => {
> +  cy.on("window:alert", (alertMessage) => {
> +    expect(alertMessage).to.eq(err);
> +  });
> +  if (typeVal) cy.get(selector).type(typeVal);
> +  cy.get(SELECTORS.CAR_NAME_FORM).submit();
> +};
> +
> +export const submitForm = (selector, typeVal) => {
> +  if (typeVal) cy.get(selector).type(typeVal);
> +  cy.get(SELECTORS.CAR_NAME_FORM).submit();
> +};
>
> 이 친구들은 Custom Commands를 활용해보셔도 좋을거 같네요
> ------------------------------
>
> In index.html
> <https://github.com/next-step/js-racingcar/pull/101#discussion_r928912002>
> :
>
> > +        <form id="trial-num__form">
> +          <fieldset id="trial-num" class="hidden">
>              <p>시도할 횟수를 입력해주세요.</p>
>              <div class="d-flex">
> -              <input type="number" class="w-100 mr-2" placeholder="시도 횟수" />
> -              <button type="button" class="btn btn-cyan">확인</button>
> +              <input
> +                id="trial-num__input"
> +                type="number"
> +                class="w-100 mr-2"
> +                placeholder="시도 횟수"
> +              />
> +              <button id="trial-num__btn" type="submit" class="btn btn-cyan">
> +                확인
> +              </button>
>              </div>
>            </fieldset>
>          </form>
>
> 👍👍👍
> ------------------------------
>
> In src/index.js
> <https://github.com/next-step/js-racingcar/pull/101#discussion_r928912753>
> :
>
> > @@ -0,0 +1,44 @@
> +import { SELECTORS } from "./constants.js";
> +import { $ } from "./dom.js";
> +import { startRacingGame } from "./racing.js";
> +import { validateNameLength, validateNumRange } from "./validation.js";
> +import { createTemplateCarPlayer } from "./template.js";
> +import { displaySelector } from "./utils.js";
> +
> +const handleSubmitCarName = (e) => {
> +  e.preventDefault();
> +
> +  // 각 자동차 이름이 5글자가 넘으면 경고 메세지를 표시한다.
> +  if (e.target[1].value) {
>
> 1과 같은 매직넘버를 사용하지 않고 값을 가져올 수 있는 방법을 찾아보시면 좋을거 같습니다
> ------------------------------
>
> In src/index.js
> <https://github.com/next-step/js-racingcar/pull/101#discussion_r928913449>
> :
>
> > +
> +const handleSubmitCarName = (e) => {
> +  e.preventDefault();
> +
> +  // 각 자동차 이름이 5글자가 넘으면 경고 메세지를 표시한다.
> +  if (e.target[1].value) {
> +    e.target[1].value.split(",").map((name) => {
> +      try {
> +        validateNameLength(name);
> +      } catch (error) {
> +        alert(error);
> +        return;
> +      }
> +      createTemplateCarPlayer($(SELECTORS.CAR_NAME_INPUT));
> +      displaySelector($(SELECTORS.TRIAL_NUM_FIELDSET));
> +      $(SELECTORS.TRIAL_NUM_INPUT).focus();
>
> focus() 넣어 주신 부분 좋네요! 👍 👍 👍
> ------------------------------
>
> In src/index.js
> <https://github.com/next-step/js-racingcar/pull/101#discussion_r928916168>
> :
>
> > +const handleSubmitCarName = (e) => {
> +  e.preventDefault();
> +
> +  // 각 자동차 이름이 5글자가 넘으면 경고 메세지를 표시한다.
> +  if (e.target[1].value) {
> +    e.target[1].value.split(",").map((name) => {
> +      try {
> +        validateNameLength(name);
> +      } catch (error) {
> +        alert(error);
> +        return;
> +      }
> +      createTemplateCarPlayer($(SELECTORS.CAR_NAME_INPUT));
> +      displaySelector($(SELECTORS.TRIAL_NUM_FIELDSET));
> +      $(SELECTORS.TRIAL_NUM_INPUT).focus();
> +    });
> +  }
> +};
>
> guard문을 위로 올려서 early return 해주시면 코드 보기가 편해질거 같네요
> ⬇️ Suggested change
>
> -const handleSubmitCarName = (e) => {
> -  e.preventDefault();
> -
> -  // 각 자동차 이름이 5글자가 넘으면 경고 메세지를 표시한다.
> -  if (e.target[1].value) {
> -    e.target[1].value.split(",").map((name) => {
> -      try {
> -        validateNameLength(name);
> -      } catch (error) {
> -        alert(error);
> -        return;
> -      }
> -      createTemplateCarPlayer($(SELECTORS.CAR_NAME_INPUT));
> -      displaySelector($(SELECTORS.TRIAL_NUM_FIELDSET));
> -      $(SELECTORS.TRIAL_NUM_INPUT).focus();
> -    });
> -  }
> -};
> +const handleSubmitCarName = (e) => {
> +  e.preventDefault();
> +
> +  if (!e.target[1].value) {
> +    return;
> +  }
> +
> +  e.target[1].value.split(",").map((name) => {
> +    try {
> +      validateNameLength(name);
> +    } catch (error) {
> +      alert(error);
> +      return;
> +    }
> +    createTemplateCarPlayer($(SELECTORS.CAR_NAME_INPUT));
> +    displaySelector($(SELECTORS.TRIAL_NUM_FIELDSET));
> +    $(SELECTORS.TRIAL_NUM_INPUT).focus();
> +  });
> +};
>
> ------------------------------
>
> In src/index.js
> <https://github.com/next-step/js-racingcar/pull/101#discussion_r928917084>
> :
>
> > @@ -0,0 +1,44 @@
> +import { SELECTORS } from "./constants.js";
> +import { $ } from "./dom.js";
> +import { startRacingGame } from "./racing.js";
> +import { validateNameLength, validateNumRange } from "./validation.js";
> +import { createTemplateCarPlayer } from "./template.js";
> +import { displaySelector } from "./utils.js";
> +
> +const handleSubmitCarName = (e) => {
> +  e.preventDefault();
> +
> +  // 각 자동차 이름이 5글자가 넘으면 경고 메세지를 표시한다.
>
> 반드시 필요한 주석이 아니라면 지워주셔도 좋을거 같네요
> ------------------------------
>
> In src/validation.js
> <https://github.com/next-step/js-racingcar/pull/101#discussion_r928918526>
> :
>
> > +export const validateNameLength = ($name) => {
> +  if ($name.length > 5) {
> +    throw ERROR_MESSAGES.WORD_LENGTH_ERROR;
> +  }
> +};
> +
> +export const validateNumRange = ($num) => {
> +  if ($num > 10 || $num < 1) {
> +    throw ERROR_MESSAGES.NUM_RANGE_ERROR;
> +  }
> +};
>
> ERROR_MESSAGES가 import 되지 않아서 에러메세지가 아래 스크린샷처럼 출력되네요
>
> [image: image]
> <https://user-images.githubusercontent.com/61097373/180795207-fc72c685-5aa5-4556-b885-3ec82fe7e85b.png>
> ------------------------------
>
> In src/template.js
> <https://github.com/next-step/js-racingcar/pull/101#discussion_r928920577>
> :
>
> > +export const templateSpinner = `<div class="spinners d-flex justify-center mt-3">
> +                                  <div class="relative spinner-container">
> +                                    <span class="material spinner"></span>
> +                                  </div>
> +                                </div>`;
> +
> +export const templateForward = `<div class="forward-icon mt-2">⬇️️</div>`;
> +
> +export const createTemplateCarPlayer = ($input) => {
> +  const template = $input.value.split(",").map(
> +    (name) => `<div class="car">
> +                  <div class="car-player">${name}</div>
> +                  <div class="spinners d-flex justify-center mt-3">
> +                    <div class="relative spinner-container">
> +                      <span class="material spinner"></span>
> +                    </div>
> +                  </div>
> +                </div>`
> +  );
> +
> +  $(SELECTORS.CAR_PLAYER_WRAPPER_DIV).innerHTML = template;
> +};
>
> templateSpinner와 templateForward는 단순히 템플릿을 저장하는 변수의 역할만 하고 있는데
> createTemplateCarPlayer는 템플릿을 생성해서 DOM에 추가해주는 역할까지 하고 있네요.
>
> createTemplateCarPlayer가 하는 두 가지 일을 다른 함수로 분리해보시면 어떨까요?
> ------------------------------
>
> In src/utils.js
> <https://github.com/next-step/js-racingcar/pull/101#discussion_r928922863>
> :
>
> > +export const displaySelector = ($selector) => {
> +  $selector.classList.remove("hidden");
> +};
>
> displaySelector라는 이름은 display를 select하는 함수라고 생각이 될 수도 있을거 같아요
>
> 다른 이름으로 바꿔보는건 어떨까요?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/next-step/js-racingcar/pull/101#pullrequestreview-1049388196>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM7W2LAR4BZ3ASEJXG5DTGTVV2O3BANCNFSM54Q52VBA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
